File: ships-backend/app/services/intelligence/dependency_analyzer.py
"""
Dependency Analyzer

Generates dependency_graph.json using dependency-cruiser for module-level dependencies.
Also provides circular dependency detection and orphaned file detection.
"""

import logging
import subprocess
import json
from pathlib import Path
from typing import Dict, Any, List, Optional
from datetime import datetime, timezone

logger = logging.getLogger("ships.intelligence")


class DependencyAnalyzer:
    """
    Generates module-level dependency graphs.
    
    Uses dependency-cruiser (Node.js) for comprehensive analysis,
    with fallback to import parsing for projects without npm.
    """
    
    def __init__(self):
        self.cache: Dict[str, Any] = {}
    
    def analyze_project(self, project_path: str) -> Dict[str, Any]:
        """
        Analyze project dependencies.
        
        Args:
            project_path: Root directory of the project
            
        Returns:
            Dependency graph with nodes, edges, circular deps
        """
        root = Path(project_path)
        
        # Check if this is a Node.js project
        has_package_json = (root / "package.json").exists()
        
        if has_package_json:
            result = self._run_dependency_cruiser(root)
            if result.get("success"):
                return result
        
        # Fallback to manual analysis
        return self._analyze_manual(root)
    
    def _run_dependency_cruiser(self, root: Path) -> Dict[str, Any]:
        """Run dependency-cruiser for comprehensive analysis."""
        try:
            # Find source directory
            src_dirs = ["src", "app", "lib"]
            source_dir = "."
            for src in src_dirs:
                if (root / src).exists():
                    source_dir = src
                    break
            
            logger.info(f"[DEP_ANALYZER] Running dependency-cruiser on {source_dir}")
            
            cmd = [
                "npx", "-y", "dependency-cruiser",
                "--output-type", "json",
                source_dir
            ]
            
            result = subprocess.run(
                cmd,
                cwd=str(root),
                capture_output=True,
                text=True,
                timeout=120
            )
            
            if result.returncode != 0:
                logger.warning(f"[DEP_ANALYZER] dependency-cruiser failed: {result.stderr[:200]}")
                return {"success": False, "error": result.stderr}
            
            data = json.loads(result.stdout)
            
            return self._transform_cruiser_output(data)
            
        except subprocess.TimeoutExpired:
            return {"success": False, "error": "Timeout"}
        except json.JSONDecodeError as e:
            return {"success": False, "error": f"Invalid JSON: {e}"}
        except Exception as e:
            return {"success": False, "error": str(e)}
    
    def _transform_cruiser_output(self, data: Dict) -> Dict[str, Any]:
        """Transform dependency-cruiser output to our format."""
        nodes = {}
        edges = []
        external_deps = set()
        
        for module in data.get("modules", []):
            source = module.get("source", "")
            
            imports = []
            imported_by = []  # Populated later
            
            for dep in module.get("dependencies", []):
                resolved = dep.get("resolved", dep.get("module", ""))
                is_external = dep.get("coreModule") or dep.get("externalPackage")
                
                if is_external:
                    external_deps.add(resolved)
                
                imports.append(resolved)
                
                edges.append({
                    "from": source,
                    "to": resolved,
                    "type": "import",
                    "is_external": is_external
                })
            
            nodes[source] = {
                "type": "file",
                "imports": imports,
                "imported_by": [],
                "external_deps": list(external_deps)
            }
        
        # Populate imported_by
        for edge in edges:
            if edge["to"] in nodes and not edge["is_external"]:
                nodes[edge["to"]]["imported_by"].append(edge["from"])
        
        # Find circular dependencies
        circular = self._detect_cycles(nodes)
        
        # Find orphaned files (not imported by anyone)
        orphaned = [
            path for path, node in nodes.items()
            if not node["imported_by"] and not path.endswith("index.ts") and not path.endswith("index.js")
        ]
        
        # Find entry points
        entry_points = self._detect_entry_points(nodes)
        
        return {
            "version": "1.0",
            "generated_at": datetime.now(timezone.utc).isoformat(),
            "success": True,
            "analyzer": "dependency-cruiser",
            "nodes": nodes,
            "edges": edges,
            "circular_dependencies": circular,
            "orphaned_files": orphaned[:10],  # Limit to 10
            "entry_points": entry_points,
            "external_dependencies": list(external_deps),
            "stats": {
                "total_modules": len(nodes),
                "total_edges": len(edges),
                "circular_count": len(circular),
                "orphaned_count": len(orphaned),
                "external_dep_count": len(external_deps)
            }
        }
    
    def _detect_cycles(self, nodes: Dict[str, Any]) -> List[List[str]]:
        """Detect circular dependencies using DFS."""
        visited = set()
        rec_stack = set()
        cycles = []
        
        def dfs(node: str, path: List[str]) -> None:
            visited.add(node)
            rec_stack.add(node)
            
            if node not in nodes:
                rec_stack.discard(node)
                return
            
            for dep in nodes[node].get("imports", []):
                if dep not in visited:
                    dfs(dep, path + [dep])
                elif dep in rec_stack:
                    try:
                        cycle_start = path.index(dep)
                        cycle = path[cycle_start:] + [dep]
                        if len(cycle) > 1 and cycle not in cycles:
                            cycles.append(cycle)
                    except ValueError:
                        pass
            
            rec_stack.discard(node)
        
        for module in nodes:
            if module not in visited:
                dfs(module, [module])
        
        return cycles[:10]  # Limit to 10 cycles
    
    def _detect_entry_points(self, nodes: Dict[str, Any]) -> List[str]:
        """Detect likely entry points."""
        entry_patterns = ["index", "main", "app", "server", "entry"]
        
        entry_points = []
        for path in nodes:
            name = Path(path).stem.lower()
            if any(pattern in name for pattern in entry_patterns):
                entry_points.append(path)
        
        return entry_points[:5]  # Limit to 5
    
    def _analyze_manual(self, root: Path) -> Dict[str, Any]:
        """Manual analysis using import parsing."""
        import re
        
        nodes = {}
        edges = []
        
        extensions = ["*.py", "*.ts", "*.tsx", "*.js", "*.jsx"]
        ignore_dirs = {"node_modules", ".venv", "venv", "__pycache__", "dist", "build"}
        
        for ext in extensions:
            for file_path in root.rglob(ext):
                if any(ignore in str(file_path) for ignore in ignore_dirs):
                    continue
                
                try:
                    content = file_path.read_text(encoding="utf-8")
                    rel_path = str(file_path.relative_to(root))
                    
                    # Find imports
                    imports = []
                    
                    # Python imports
                    if file_path.suffix == ".py":
                        import_pattern = r"^(?:from\s+(\S+)|import\s+(\S+))"
                        for match in re.finditer(import_pattern, content, re.MULTILINE):
                            module = match.group(1) or match.group(2)
                            if module:
                                imports.append(module.split(".")[0])
                    
                    # JS/TS imports
                    else:
                        import_pattern = r"import\s+.*?from\s+['\"](.+?)['\"]"
                        require_pattern = r"require\s*\(['\"](.+?)['\"]\)"
                        
                        imports.extend(re.findall(import_pattern, content))
                        imports.extend(re.findall(require_pattern, content))
                    
                    nodes[rel_path] = {
                        "type": "file",
                        "imports": imports,
                        "imported_by": []
                    }
                    
                except Exception as e:
                    logger.warning(f"[DEP_ANALYZER] Failed to parse {file_path}: {e}")
        
        return {
            "version": "1.0",
            "generated_at": datetime.now(timezone.utc).isoformat(),
            "success": True,
            "analyzer": "manual",
            "nodes": nodes,
            "edges": edges,
            "circular_dependencies": [],
            "orphaned_files": [],
            "entry_points": [],
            "stats": {
                "total_modules": len(nodes)
            }
        }
    
    def save_artifact(self, project_path: str, output_path: Optional[str] = None) -> str:
        """Analyze project and save to dependency_graph.json."""
        result = self.analyze_project(project_path)
        
        if not output_path:
            output_path = Path(project_path) / ".ships" / "dependency_graph.json"
        else:
            output_path = Path(output_path)
        
        output_path.parent.mkdir(parents=True, exist_ok=True)
        
        with open(output_path, "w", encoding="utf-8") as f:
            json.dump(result, f, indent=2)
        
        logger.info(f"[DEP_ANALYZER] Saved dependency_graph.json to {output_path}")
        
        return str(output_path)

File: ships-preview/src/main/services/dependencyAnalyzer.ts
/**
 * Dependency Analyzer - Production Grade
 *
 * Uses dependency-cruiser to analyze module dependencies.
 * Detects circular dependencies and orphaned files.
 */

import * as fs from 'fs';
import * as path from 'path';
import { exec } from 'child_process';
import { promisify } from 'util';

const execAsync = promisify(exec);

export interface DependencyNode {
  source: string;
  imports: string[];
  importedBy: string[];
  externalDeps: string[];
  coreNodeModules: string[];
}

export interface CircularDependency {
  cycle: string[];
  length: number;
}

export interface DependencyGraph {
  version: string;
  generatedAt: string;
  projectPath: string;
  totalModules: number;
  nodes: Record<string, DependencyNode>;
  circularDependencies: CircularDependency[];
  orphanedFiles: string[];
  externalPackages: string[];
}

/**
 * Production dependency analyzer using dependency-cruiser
 */
export class DependencyAnalyzer {
  private projectPath: string;

  constructor(projectPath: string) {
    this.projectPath = projectPath;
  }

  /**
   * Generate dependency_graph.json
   */
  async generateDependencyGraph(): Promise<DependencyGraph> {
    let nodes: Record<string, DependencyNode> = {};
    let circularDependencies: CircularDependency[] = [];
    const externalPackages = new Set<string>();

    // Try dependency-cruiser first
    try {
      const result = await this.runDependencyCruiser();
      nodes = result.nodes;
      circularDependencies = result.circular;
      result.external.forEach(pkg => externalPackages.add(pkg));
    } catch (error) {
      console.warn('[DependencyAnalyzer] dependency-cruiser failed, using fallback:', error);
      nodes = await this.analyzeFallback();
    }

    // Detect orphaned files
    const orphanedFiles = Object.entries(nodes)
      .filter(([_, node]) => node.importedBy.length === 0 && !node.source.includes('index'))
      .map(([source]) => source)
      .slice(0, 20);

    const graph: DependencyGraph = {
      version: '1.0.0',
      generatedAt: new Date().toISOString(),
      projectPath: this.projectPath,
      totalModules: Object.keys(nodes).length,
      nodes,
      circularDependencies,
      orphanedFiles,
      externalPackages: Array.from(externalPackages),
    };

    // Write to .ships/
    const shipsDir = path.join(this.projectPath, '.ships');
    if (!fs.existsSync(shipsDir)) {
      fs.mkdirSync(shipsDir, { recursive: true });
    }
    fs.writeFileSync(
      path.join(shipsDir, 'dependency_graph.json'),
      JSON.stringify(graph, null, 2)
    );

    return graph;
  }

  /**
   * Run dependency-cruiser via npx
   */
  private async runDependencyCruiser(): Promise<{
    nodes: Record<string, DependencyNode>;
    circular: CircularDependency[];
    external: string[];
  }> {
    const nodes: Record<string, DependencyNode> = {};
    const circular: CircularDependency[] = [];
    const external: string[] = [];

    // Find source directories
    const srcDir = fs.existsSync(path.join(this.projectPath, 'src')) ? 'src' : '.';
    
    const { stdout } = await execAsync(
      `npx -y dependency-cruiser --output-type json ${srcDir}`,
      { 
        cwd: this.projectPath, 
        timeout: 120000,
        maxBuffer: 10 * 1024 * 1024 // 10MB
      }
    );

    const data = JSON.parse(stdout);

    // Process modules
    for (const module of data.modules || []) {
      const source = module.source;
      
      nodes[source] = {
        source,
        imports: [],
        importedBy: [],
        externalDeps: [],
        coreNodeModules: [],
      };

      for (const dep of module.dependencies || []) {
        const resolved = dep.resolved || dep.module;
        
        if (dep.dependencyTypes?.includes('npm')) {
          nodes[source].externalDeps.push(dep.module);
          external.push(dep.module.split('/')[0]);
        } else if (dep.dependencyTypes?.includes('core')) {
          nodes[source].coreNodeModules.push(dep.module);
        } else if (resolved) {
          nodes[source].imports.push(resolved);
        }

        // Check for circular
        if (dep.circular) {
          circular.push({
            cycle: dep.cycle || [source, resolved],
            length: dep.cycle?.length || 2,
          });
        }
      }
    }

    // Populate importedBy
    for (const [source, node] of Object.entries(nodes)) {
      for (const imp of node.imports) {
        if (nodes[imp]) {
          nodes[imp].importedBy.push(source);
        }
      }
    }

    return { nodes, circular, external };
  }

  /**
   * Fallback: Basic import parsing
   */
  private async analyzeFallback(): Promise<Record<string, DependencyNode>> {
    const nodes: Record<string, DependencyNode> = {};

    const scanDir = async (dir: string): Promise<void> => {
      const entries = fs.readdirSync(dir, { withFileTypes: true });

      for (const entry of entries) {
        if (entry.name === 'node_modules' || entry.name.startsWith('.')) continue;

        const fullPath = path.join(dir, entry.name);

        if (entry.isDirectory()) {
          await scanDir(fullPath);
        } else if (entry.isFile() && /\.(ts|tsx|js|jsx|py)$/.test(entry.name)) {
          const relativePath = path.relative(this.projectPath, fullPath).replace(/\\/g, '/');
          const content = fs.readFileSync(fullPath, 'utf-8');
          const imports = this.extractImports(content, entry.name);

          nodes[relativePath] = {
            source: relativePath,
            imports: imports.internal,
            importedBy: [],
            externalDeps: imports.external,
            coreNodeModules: imports.core,
          };
        }
      }
    };

    await scanDir(this.projectPath);

    // Populate importedBy
    for (const [source, node] of Object.entries(nodes)) {
      for (const imp of node.imports) {
        if (nodes[imp]) {
          nodes[imp].importedBy.push(source);
        }
      }
    }

    return nodes;
  }

  /**
   * Extract imports from file content
   */
  private extractImports(content: string, filename: string): {
    internal: string[];
    external: string[];
    core: string[];
  } {
    const internal: string[] = [];
    const external: string[] = [];
    const core: string[] = [];

    const coreModules = new Set([
      'fs', 'path', 'http', 'https', 'crypto', 'os', 'child_process',
      'util', 'events', 'stream', 'buffer', 'url', 'querystring'
    ]);

    if (filename.endsWith('.py')) {
      // Python imports
      const regex = /^(?:from\s+(\S+)|import\s+(\S+))/gm;
      let match;
      while ((match = regex.exec(content)) !== null) {
        const module = match[1] || match[2];
        if (module.startsWith('.')) {
          internal.push(module);
        } else {
          external.push(module.split('.')[0]);
        }
      }
    } else {
      // JS/TS imports
      const regex = /import\s+(?:{[^}]+}|\w+|\*\s+as\s+\w+)\s+from\s+['"]([^'"]+)['"]/g;
      let match;
      while ((match = regex.exec(content)) !== null) {
        const module = match[1];
        if (module.startsWith('.')) {
          internal.push(module);
        } else if (coreModules.has(module)) {
          core.push(module);
        } else {
          external.push(module.split('/')[0]);
        }
      }

      // require() calls
      const reqRegex = /require\(['"]([^'"]+)['"]\)/g;
      while ((match = reqRegex.exec(content)) !== null) {
        const module = match[1];
        if (module.startsWith('.')) {
          internal.push(module);
        } else if (coreModules.has(module)) {
          core.push(module);
        } else {
          external.push(module.split('/')[0]);
        }
      }
    }

    return { internal, external: [...new Set(external)], core: [...new Set(core)] };
  }
}
